design-proposals/mass_machine_type_transition.md

```diff
@@ -0,0 +1,208 @@
+
+## Non Goals
+While these additions could be beneficial to cluster-admins who wish to use the mass machine type transition (MMTT), these are not covered in the scope of this design proposal: 
+* Allowing user to filter VMs by running state: a cluster admin may want to only update VMs that are offline.
+* Allowing user to filter VMs in ways other than by namespace; e.g. with label-selector
+
+Also not covered in this design proposal is the method in which users/cluster admins will be informed that they have VMs with unsupported machine types.
+
+## User Stories
+As a cluster-admin, I want an automated way to update the machine type of all VMs to be compatible with RHEL 10 without interrupting my workflow.
+As a cluster-admin, I want to be able to include only certain VMs to be updated (e.g. by namespace).
+As a cluster-admin, I want to be in control of the behavior of running VMs being updated; I can determine whether I want to manually restart running VMs being updated or let the automation restart all of them for me. 
+
+## Repos
+[kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+Create a new virtctl command that will allow the user to automatically update the machine type of any VMs with a machine type that is no longer supported. This command invokes a k8s `Job` that iterates through all VMs within a specified namespace (or all namespaces if none is specified), determines if the machine type is no longer supported, and updates it to the latest supported version. 
+
+If the VM is running, the update will not take effect until it is restarted, so the label `restart-vm-required` will be applied to the VM until the VMI is removed (indicating the VM has been stopped), upon which the label will be cleared. The job will only terminate once all `restart-vm-required` labels have been cleared. By default, the user must handle stopping/restarting the labelled VMs manually. This is to allow the user to safely update the machine types of their VMs at a time they choose without worrying about workload interruptions. However, the user also has the option to allow the job to automatically restart every running VM that is being updated and apply the changes immediately.
+
+The options to select a specific namespace or restart the VMs immediately are configurable with the `NAMESPACE` and `RESTART_NOW` environment variables respectively.
+
+Given a VM with the following spec:
+````
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: false
+  template:
+    spec:
+      domain:
+        machine:
+          type: pc-q35-rhel8.2.0
+````
+When the MMTT is invoked, the VM spec will be updated to: 
+````
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: false
+  template:
+    spec:
+      domain:
+        machine:
+          type: pc-q35-rhel9.2.0
+````
+where `pc-q35-rhel9.2.0` is the latest machine type version.
```

Comment:
I guess the job is using `/usr/libexec/qemu-kvm -machine ?` to determine this?

Could we also consider the target machine type being provided by the admin?

Reply:
In the case of allowing the admin configure the target machine type, do we want to limit it to only supported machine types (in this case, just the RHEL-9 machine types)?

Reply:
I think it would be ideal that the admin would choose between a specified type and the `latest` type that would be determined dynamically. Perhaps we can utilize `node-labeller` to populate the latest machine type supported, then we could fetch the latest version that's supported on all nodes.

> do we want to limit it to only supported machine types (in this case, just the RHEL-9 machine types)?

We could, via `/usr/libexec/qemu-kvm -machine ?`, parse the different versions to give the admin a choice, but I would avoid providing specific hard-coded versions that would demand us updating the versions manually whenever needed.

Reply:
> I think it would be ideal that the admin would choose between a specified type and the `latest` type that would be determined dynamically. Perhaps we can utilize `node-labeller` to populate the latest machine type supported, then we could fetch the latest version that's supported on all nodes.

They can already use the aliases to track the `latest` machine type, you could have another way for them to pick and pin to the latest machine type but that might be over complicating things a little.

> > do we want to limit it to only supported machine types (in this case, just the RHEL-9 machine types)?
> 
> We could, via `/usr/libexec/qemu-kvm -machine ?`, parse the different versions to give the admin a choice, but I would avoid providing specific hard-coded versions that would demand us updating the versions manually whenever needed.

Getting the supported versions dynamically might be awkward tbh, in the past for OpenStack I just ended up defining some basic regex covering the known type names (note we had to handle upstream *and* downstream QEMU while AFAIK KubeVirt only supports downstream CentOS Stream QEMU right?).

https://github.com/openstack/nova/blob/4b454febf73cdd7b5be0a2dad272c1d7685fac9e/nova/virt/libvirt/machine_type_utils.py#L24-L39

Reply:
> the latest type that would be determined dynamically

Would replacing the machine types with just `q35` and restarting the VMs be sufficient then for updating the machine types to the latest version, since 'q35' is the alias for the latest version? Then if the admin wishes to use a different target version they can specify (choosing from a list of supported machine types? or specifying any machine type?)

Reply:
No. `q35` is a "floating" type and iwll always "point" to the latest q35.
IIUIC then we rather want to use the most recent specific (non-floating) machine type.
Ths is required in order to then keep the machine type stable until the next update of a machine type.